src/Serialization/main.jl

```diff
@@ -73,36 +92,117 @@ function decodeType(input::String)
+exists. If `Vector{T}` is serialized with `is_basic_serialization_type(T) = true`
+then the `entry_type` keyword is used to store the type `T` as a property of
+the vector.
+
+# Examples
+
+```jldoctest
+julia> is_basic_serialization_type(ZZRingElem)
+true
+```
+"""
+is_basic_serialization_type(::Type) = false
+is_basic_serialization_type(::Type{ZZRingElem}) = true
+is_basic_serialization_type(::Type{QQFieldElem}) = true
+is_basic_serialization_type(::Type{Bool}) = true
+is_basic_serialization_type(::Type{Symbol}) = true
+# this deals with int32, int64 etc.
+is_basic_serialization_type(::Type{T}) where T <: Number = isconcretetype(T)
+
+# ATTENTION
+# We need to distinguish between data with a globally defined normal form and data where such a normal form depends on some parameters.
+# In particular, this does NOT ONLY depend on the type; see, e.g., FqField.
+function has_elem_basic_encoding(obj::T) where T <: Ring
+    return is_basic_serialization_type(elem_type(obj))
+end
+
+has_basic_encoding(obj::T) where T = is_basic_serialization_type(T)
+
+function has_basic_encoding(obj::T) where T <: RingElem
+    return has_elem_basic_encoding(parent(obj))
+end
+
+has_elem_basic_encoding(obj::T) where T = false
+has_elem_basic_encoding(obj::ArbField) = true
+has_elem_basic_encoding(obj::AcbField) = true
+has_elem_basic_encoding(obj::FqField) = absolute_degree(obj) == 1
+has_elem_basic_encoding(obj::Nemo.fpField) = true
+has_elem_basic_encoding(obj::Nemo.FpField) = true
+has_elem_basic_encoding(obj::Nemo.zzModRing) = true
+has_elem_basic_encoding(obj::FlintPadicField) = true
+has_elem_basic_encoding(obj::TropicalSemiring) = true
 
 
 ################################################################################
 # High level
-function save_type_dispatch(s::SerializerState, obj::T) where T
-    if is_basic_serialization_type(T) && s.depth != 0
-        return save_internal(s, obj)
+
+function load_ref(s::DeserializerState, dict::Dict)
+    if !haskey(dict, :id)
+        return load_unknown_type(s, dict)
     end
 
-    if !isprimitivetype(T) && !Base.issingletontype(T)
-        # if obj is already serialzed, just output
-        # a backref
-        ref = get(s.objmap, obj, nothing)
-        if ref !== nothing
-            return Dict{Symbol, Any}(
-              :type => backref_sym,
-              :id => string(ref),
-              :version => 1, # ???
-              )
-        end
-        # otherwise,
-        ref = s.objmap[obj] = uuid4()
+    id = dict[:id]
+
+    if haskey(s.objs, UUID(id))
+        loaded_ref = s.objs[UUID(id)]
     else
-        ref = nothing
+        ref_dict = s.refs[Symbol(id)]
+        ref_dict[:id] = id
+        loaded_ref = load_unknown_type(s, ref_dict)
     end
+    return loaded_ref
+end
 
-    result = Dict{Symbol, Any}(:type => encodeType(T))
+function save_as_ref(s::SerializerState, obj::T) where T
+    if has_elem_basic_encoding(obj) && s.depth != 0
```

Comment:
Maybe add a comment here. 
```suggestion
    # For rings whose elements have basic encoding (e.g. Z/7Z), we don't bother
    # storing backrefs to the ring, we just create a new instance of the ring.
    if has_elem_basic_encoding(obj) && s.depth != 0
```

Also: when can it happens that `s.depth == 0`? That would mean we invoke `save_as_ref` as the "top level", no? And anyway, how can we then store a backref if we created all those rings without an id ?!?!?

Maybe we can either give an example (in a comment, and/or a test); or else argue that it cannot happen (and then we just put in an `@assert s.depth != 0` or something like this)

Reply:
Actually, now I wonder: shouldn't this use `serialize_with_id` here? I am confused, sorry :-)

Reply:
this now checks if serialize with id is false and returns the encoded obj, storing of any  parent types now go through a save_ref, maybe this means the function deserves a name change to save_parent